src/Magnum/Octree/Octree.hpp

```diff
@@ -0,0 +1,380 @@
+                resultData.push_back(_data[i]);
+        }
+
+        if(!isLeafNode(nodeIndex)) {
+            const Int childrenStart = _childrenStart[nodeIndex];
+            for(Int i = 0; i < NodesPerLevel; ++i) {
+                points(resultData, childrenStart + i);
+            }
+        }
+    }
+
+    // For 100% accuracy additional test with the corners of the frustum is needed (planes are infinite)!
+
+    return *this;
+}
+
+template<UnsignedInt dimensions, class T, class Data>
+const Tree<dimensions, T, Data>& Tree<dimensions, T, Data>::points(std::vector<Data>& resultData, const Int nodeIndex) const {
+    /* Root node means entire tree, copy all data */
+    if(nodeIndex == 0) {
+        std::copy(_data.begin(), _data.end(), std::back_inserter(resultData));
+        return *this;
+    }
+
+    Containers::Array<Int> nodeStack{size_t(_maxDepth)};
+    Int stackFront = -1;
+    int curNode = nodeIndex;
+
+    do {
+        if(curNode != 0) {
+            nodeStack[++stackFront] = curNode;
+            while(curNode != 0) {
+                curNode = nodeStack[++stackFront] = _childrenStart[nodeIndex];
+            }
+        }
+
+        const Int poppedNode = nodeStack[stackFront--];
+        if(isLeafNode(poppedNode)) {
+            /* Shortcut: Add data of this leaf and all its siblings */
+            const auto& dataRangeFirst = _perNodeData[poppedNode];
+            const auto& dataRangeLast = _perNodeData[poppedNode+7];
+            std::copy_n(_data.prefix(dataRangeFirst[0]).data(), dataRangeLast[1] - dataRangeFirst[0], resultData.end());
+
+            curNode = 0;
+        } else {
+            /* Copy the data from this node */
+            const auto& dataRange = _perNodeData[poppedNode];
+            std::copy_n(_data.prefix(dataRange[0]).data(), dataRange[1] - dataRange[0], resultData.end());
+
+            /* Set curNode to next sibling, if not the last node */
+            // TODO make number of bits "NodesPerLevel" dependent!
+            curNode = ((poppedNode & 0b111) != 0) ? poppedNode + 1 : 0;
+        }
+    } while(curNode != 0 || stackFront > 0);
+
+    return *this;
+}
+
+template<UnsignedInt dimensions, class T, class Data>
+OctantStatus Tree<dimensions, T, Data>::cubeInFrustum(const Math::Frustum<T>& frustum, const Int nodeIndex) const {
+    /* Check for each corner of an octant whether it is inside the frustum.
+       If only some of the corners are inside, the octant requires further checks. */
+    Int planes = 0;
+
+    const Math::Vector3<T> min = _centers[nodeIndex] - Math::Vector3<T>{_radii[nodeIndex]};
+    const Math::Vector3<T> max = _centers[nodeIndex] + Math::Vector3<T>{_radii[nodeIndex]};
+
+    for(const Math::Vector4<T>& plane: frustum.planes()) {
+        int corners = 0;
+
+        for(UnsignedByte c = 0; c < 8; ++c) {
+            const Math::Vector3<T> corner = Math::lerp(min, max, Math::BoolVector<3>{c});
```

Comment:
If doing intersection test between an AABB and a Frustum, only 2 vertices is needed. This is an optimization you can consider.
For technical details, you can read the chapter, `Geometric Approach - Testing Boxes II` in this [article](https://cgvr.cs.uni-bremen.de/teaching/cg_literatur/lighthouse3d_view_frustum_culling/index.html).

Reply:
Or you can directly use `aabbFrustum` from `Magnum\Math\Intersection.h`

Reply:
Hi @bigbike!

Currently neither @ACapo nor I am actively working on this PR, especially since we figured out that using `aabbFrustum` (which I contributed btw :) https://github.com/mosra/magnum/commit/04ca48ce649625a2eebd1c11a7ed5ca37ddc75e4) directly was just as fast for our use case (we only had around 60-80 objects) and building the octree actually caused more overhead...

This is still in the back of my mind and I really want to finish it, but at the moment I can't take the time.

Best, Jonathan.

Reply:
Hi @Squareys ,

Good to see your reply!
Before commenting yesterday, I read the source code of `aabbFrustum`, which was implemented in a very elegant, efficient manner. A pieces of very high-quality code. Did not know the author was you. I like it a lot. :-)

I like the code in this PR as well. Do not give up! Such functionality is rather necessary and important especially when rendering large-scale, static, geometrically-complex models.

Nice meeting you in this PR. Thanks!

Yili